<a href="https://colab.research.google.com/github/wiv33/A-Learning-python/blob/master/machine-learning/_000_hello_machine/_000_basic/_003_tensorflow_cuk_edu/_002_2_ml/_002_weather_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 기상청 날씨 데이터 분석하기

## 데이터 불러오기

- 컬럼 설정
- 사람을 위한 월 시작 row 제거
  - 12월 31일 -> | start: 201912 | -> 1월 1일

In [162]:
import pandas as pd
path = 'https://raw.githubusercontent.com/wiv33/A-Learning-python/master/machine-learning/_000_hello_machine/_000_basic/_003_tensorflow_cuk_edu/_002_2_ml/data'

df_precipitation = pd.read_csv(path + '/precipitation_201910_202009.csv')
df_precipitation.columns = ['day', 'hour', 'precipitation']
df_precipitation.dropna(axis=0, inplace=True)

df_direction_of_the_wind = pd.read_csv(path + '/direction_of_the_wind_201910_202009.csv')
df_direction_of_the_wind.columns = ['day', 'hour', 'wind']
df_direction_of_the_wind.dropna(axis=0, inplace=True)

df_precipitation_shape = pd.read_csv(path + '/precipitation_shape_201910_202009.csv')
df_precipitation_shape.columns = ['day', 'hour', 'precipitation_shape']
df_precipitation_shape.dropna(axis=0, inplace=True)

df_temperature = pd.read_csv(path + '/temperature_201910_202009.csv')
df_temperature.columns = ['day', 'hour', 'temperature']
df_temperature.dropna(axis=0, inplace=True)

df_humidity = pd.read_csv(path + '/humidity_201910_202009.csv')
df_humidity.columns = ['day', 'hour', 'humidity']
df_humidity.dropna(axis=0, inplace=True)

df_wind_spped = pd.read_csv(path + '/%EC%9B%94%EA%B3%A1%EC%A0%9C1%EB%8F%99_%ED%92%8D%EC%86%8D_wind_speed_201910_202009.csv')
df_wind_spped.columns = ['day', 'hour', 'wind_speed']
df_wind_spped.dropna(axis=0, inplace=True)


nan row 확인

In [163]:
df_test = pd.read_csv(path + '/temperature_201910_202009.csv')
df_test.iloc[743:746, :]

format: day    hour  value location:61_128 Start : 20191001 
743                  31  2300.0                                     13.0
744   Start : 20191101      NaN                                      NaN
745                   1     0.0                                     15.3

nan 형태 확인 후 제거

In [164]:
del df_test

### 다섯 개의 DataFrame shape 확인하기

In [165]:
df_direction_of_the_wind.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   day     8784 non-null   object 
 1   hour    8784 non-null   float64
 2   wind    8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


In [166]:
df_precipitation_shape.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   day                  8784 non-null   object 
 1   hour                 8784 non-null   float64
 2   precipitation_shape  8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


In [167]:
df_humidity.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   day       8784 non-null   object 
 1   hour      8784 non-null   float64
 2   humidity  8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


In [168]:
df_temperature.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   day          8784 non-null   object 
 1   hour         8784 non-null   float64
 2   temperature  8784 non-null   float64
dtypes: float64(2), object(1)
memory usage: 274.5+ KB


8784 row와 3개의 컬럼 중 세 번째 컬럼을 하나로 만들면 될 것.

## 각 데이터 프레임 하나로 합치기

강수, 강수형태, 습도, 기온, 풍향, 풍속

In [169]:
df = df_precipitation.iloc[:, :2]
df['precipitation_강수'] = df_precipitation.precipitation
df['precipitation_shape_강수형태'] = df_precipitation_shape['precipitation_shape']
df['temperature_기온'] = df_temperature['temperature']
df['humidity_습도'] = df_humidity.humidity
df['wind_풍향'] = df_direction_of_the_wind.wind
df['wind_spped_풍속'] = df_wind_spped.wind_speed


### merge 결과 확인하기

In [170]:
df.head()

day   hour  precipitation_강수  ...  humidity_습도  wind_풍향  wind_spped_풍속
0   1    0.0               0.0  ...         63.0     29.0            0.8
1   1  100.0               0.0  ...         60.0    183.0            1.6
2   1  200.0               0.0  ...         56.0    122.0            1.6
3   1  300.0               0.0  ...         57.0     64.0            1.9
4   1  400.0               0.0  ...         59.0    127.0            1.5

[5 rows x 8 columns]

## 시간 데이터 타입 변환

- 기존 24시의 표현 변경

  0.0, 100.0, 200.0

  0, 1, 2, 3

- day int로 변경


In [171]:
def hour_regulation(hour: float):
  return int(hour / 100)

In [172]:
df.hour = df.hour.apply(hour_regulation)
df.day = df.day.apply(lambda x: int(x))
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8784 entries, 0 to 8794
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   day                       8784 non-null   int64  
 1   hour                      8784 non-null   int64  
 2   precipitation_강수          8784 non-null   float64
 3   precipitation_shape_강수형태  8784 non-null   float64
 4   temperature_기온            8784 non-null   float64
 5   humidity_습도               8784 non-null   float64
 6   wind_풍향                   8784 non-null   float64
 7   wind_spped_풍속             8784 non-null   float64
dtypes: float64(6), int64(2)
memory usage: 617.6 KB


In [173]:
df.describe()

day         hour  ...      wind_풍향  wind_spped_풍속
count  8784.000000  8784.000000  ...  8784.000000    8784.000000
mean     15.756831    11.500000  ...   204.206398       1.899841
std       8.812031     6.922581  ...   120.910504       1.220649
min       1.000000     0.000000  ...    -1.000000      -1.000000
25%       8.000000     5.750000  ...    82.000000       1.000000
50%      16.000000    11.500000  ...   239.000000       1.600000
75%      23.000000    17.250000  ...   314.000000       2.600000
max      31.000000    23.000000  ...   360.000000      12.800000

[8 rows x 8 columns]

### 이상 데이터 확인

강수, 강수 형태, 습도, 풍향, 풍속의 최소 값이 `-1`이고

기온이 `-50도`가 정상인 것 같지 않다.

In [175]:
df[df['temperature_기온'] == -50].index

day  hour  precipitation_강수  ...  humidity_습도  wind_풍향  wind_spped_풍속
514    22    10              -1.0  ...         -1.0     -1.0           -1.0
1057   14     0              -1.0  ...         -1.0     -1.0           -1.0
2271    3    12              -1.0  ...         -1.0     -1.0           -1.0
3468   22     8              -1.0  ...         -1.0     -1.0           -1.0
3475   22    15              -1.0  ...         -1.0     -1.0           -1.0
3481   22    21              -1.0  ...         -1.0     -1.0           -1.0
4183   23     2              -1.0  ...         -1.0     -1.0           -1.0
4661   11    23              -1.0  ...         -1.0     -1.0           -1.0
5487   16     8              -1.0  ...         -1.0     -1.0           -1.0

[9 rows x 8 columns]

- 이상 데이터의 인덱스 확인

### 이상 데이터 정규화

- 습도와 기온은 +-2일 값의 평균으로 적용
- 나머지 값은 0으로 적용

In [179]:
import numpy as np 

for x in df[df['wind_풍향'] <= -1].index:
  seq = int(x)
  df['temperature_기온'][seq] = np.mean((df.iloc[seq-2:seq, 4], df.iloc[seq+1:seq+3, 4]))
  df['humidity_습도'][seq] = np.mean((df.iloc[seq-2:seq, 5], df.iloc[seq+1:seq+3, 5]))
  df['wind_spped_풍속'][seq] = 0.0
  df['precipitation_shape_강수형태'][seq] = 0.0
  df['precipitation_강수'][seq] = 0.0
  df['wind_풍향'][seq] = 0.0


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A va

In [180]:
df[df['humidity_습도'] <= 10.0]

day  hour  precipitation_강수  ...  humidity_습도  wind_풍향  wind_spped_풍속
5291    8     4               0.0  ...         10.0    136.0            2.5

[1 rows x 8 columns]

In [181]:
df[df['temperature_기온'] <= -50]

Empty DataFrame
Columns: [day, hour, precipitation_강수, precipitation_shape_강수형태, temperature_기온, humidity_습도, wind_풍향, wind_spped_풍속]
Index: []

In [182]:
df.iloc[1055: 1057, :]

day  hour  precipitation_강수  ...  humidity_습도  wind_풍향  wind_spped_풍속
1056   13    23               0.0  ...        24.00      2.0            3.2
1057   14     0               0.0  ...        21.25      0.0            0.0

[2 rows x 8 columns]

In [183]:
df[df['wind_풍향'] <= -1]

Empty DataFrame
Columns: [day, hour, precipitation_강수, precipitation_shape_강수형태, temperature_기온, humidity_습도, wind_풍향, wind_spped_풍속]
Index: []

In [184]:
df.describe()

day         hour  ...      wind_풍향  wind_spped_풍속
count  8784.000000  8784.000000  ...  8784.000000    8784.000000
mean     15.756831    11.500000  ...   204.207536       1.900979
std       8.812031     6.922581  ...   120.908576       1.218408
min       1.000000     0.000000  ...     0.000000       0.000000
25%       8.000000     5.750000  ...    82.000000       1.000000
50%      16.000000    11.500000  ...   239.000000       1.600000
75%      23.000000    17.250000  ...   314.000000       2.600000
max      31.000000    23.000000  ...   360.000000      12.800000

[8 rows x 8 columns]